layout/FallbackQuery.js

```diff
@@ -0,0 +1,304 @@
+  }
+
+  // add borough if specified
+  if (vs.isset('input:borough')) {
+    o.bool.must.push(addSecondary(vs.var('input:borough').toString(), ['borough', 'borough_a']));
+  }
+
+  // add locality if specified
+  if (vs.isset('input:locality')) {
+    o.bool.must.push(addSecondary(vs.var('input:locality').toString(), ['locality', 'locality_a', 'localadmin', 'localadmin_a']));
+  }
+
+  // add region if specified
+  if (vs.isset('input:region')) {
+    o.bool.must.push(addSecondary(vs.var('input:region').toString(), ['region', 'region_a']));
+  }
+
+  if (vs.isset('input:country')) {
+    o.bool.must.push(addSecondary(vs.var('input:country').toString(), ['country', 'country_a']));
+  }
+
+  return o;
+
+}
+
+function addHouseNumberAndStreet(vs) {
+  var o = {
+    bool: {
+      must: [
+        {
+          term: { layer: 'address' }
+        },
+        {
+          match_phrase: {
+            'address_parts.number': vs.var('input:housenumber').toString()
+          }
+        },
+        {
+          match_phrase: {
+            'address_parts.street': vs.var('input:street').toString()
+          }
+        }
+      ]
+    }
+  };
+
+  // add neighbourhood if specified
+  if (vs.isset('input:neighbourhood')) {
+    o.bool.must.push(addSecondary(vs.var('input:neighbourhood').toString(), ['neighbourhood', 'neighbourhood_a']));
+  }
+
+  // add borough if specified
+  if (vs.isset('input:borough')) {
+    o.bool.must.push(addSecondary(vs.var('input:borough').toString(), ['borough', 'borough_a']));
+  }
+
+  // add locality if specified
+  if (vs.isset('input:locality')) {
+    o.bool.must.push(addSecondary(vs.var('input:locality').toString(), ['locality', 'locality_a', 'localadmin', 'localadmin_a']));
+  }
+
+  // add region if specified
+  if (vs.isset('input:region')) {
+    o.bool.must.push(addSecondary(vs.var('input:region').toString(), ['region', 'region_a']));
+  }
+
+  if (vs.isset('input:country')) {
+    o.bool.must.push(addSecondary(vs.var('input:country').toString(), ['country', 'country_a']));
+  }
+
+  return o;
+
+}
+
+function addNeighbourhood(vs) {
+  var o = addPrimary(vs.var('input:neighbourhood').toString(),
+            'neighbourhood', ['neighbourhood', 'neighbourhood_a'], false);
+
+  // add borough if specified
+  if (vs.isset('input:borough')) {
+    o.bool.must.push(addSecondary(vs.var('input:borough').toString(), ['borough', 'borough_a']));
+  }
+
+  // add locality if specified
+  if (vs.isset('input:locality')) {
+    o.bool.must.push(addSecondary(vs.var('input:locality').toString(), ['locality', 'locality_a', 'localadmin', 'localadmin_a']));
+  }
+
+  // add region if specified
+  if (vs.isset('input:region')) {
+    o.bool.must.push(addSecondary(vs.var('input:region').toString(), ['region', 'region_a']));
+  }
+
+  // add country if specified
+  if (vs.isset('input:country')) {
+    o.bool.must.push(addSecondary(vs.var('input:country').toString(), ['country', 'country_a']));
+  }
+
```

Comment:
this code feels very repetitive (more below). i wonder if there is a way to represent this all in a `map` of some sort. Something like...

```javascript
var mapping = {
  'input:borough': ['borough', 'borough_a'],
  'input:locality': ['locality', 'locality_a', 'localadmin', 'localadmin_a'],
  // etc...
}

function setSecondary(vs, o, key) {
  if (vs.isset(key)) {
    o.bool.must.push(addSecondary(vs.var(key).toString(), mapping[key]));
  }
}

Object.keys(mapping).forEach(setSecondary.bind(null,vs,o));
```